examples/acoustic/fwi_operators.py

```diff
@@ -20,66 +22,69 @@ class ForwardOperator(Operator):
-        source.data[:] = src.traces[:]
+    # Create the stencil by hand instead of calling numpy solve for speed purposes
+    # Simple linear solve of a u(t+dt) + b u(t) + c u(t-dt) = L for u(t+dt)
+    stencil = 1 / (2 * m + s * damp) * (
+        4 * m * u + (s * damp - 2 * m) * u.backward +
+        2 * s**2 * (laplacian + s**2 / 12 * biharmonic))
+    # Add substitutions for spacing (temporal and spatial)
+    subs = {s: dt, h: model.get_spacing()}
 
-        super(ForwardOperator, self).__init__(nt, m.shape,
-                                              stencils=Eq(u.forward, stencil),
-                                              subs=subs,
-                                              spc_border=max(spc_order, 2),
-                                              time_order=2,
-                                              forward=True,
-                                              dtype=m.dtype,
-                                              **kwargs)
+    if legacy:
+        op = Operator(nt, m.shape, stencils=Eq(u.forward, stencil), subs=subs,
+                      spc_border=max(spc_order / 2, 2), time_order=2, forward=True,
+                      dtype=m.dtype, **kwargs)
 
         # Insert source and receiver terms post-hoc
-        self.input_params += [source, source.coordinates, rec, rec.coordinates]
-        self.output_params += [rec]
-        self.propagator.time_loop_stencils_a = source.add(m, u) + rec.read(u)
-        self.propagator.add_devito_param(source)
-        self.propagator.add_devito_param(source.coordinates)
-        self.propagator.add_devito_param(rec)
-        self.propagator.add_devito_param(rec.coordinates)
+        op.input_params += [src, src.coordinates, rec, rec.coordinates]
+        op.output_params += [rec]
+        op.propagator.time_loop_stencils_a = src.add(m, u) + rec.read(u)
+        op.propagator.add_devito_param(src)
+        op.propagator.add_devito_param(src.coordinates)
+        op.propagator.add_devito_param(rec)
+        op.propagator.add_devito_param(rec.coordinates)
+
+    else:
+        # Create stencil expressions for operator, source and receivers
+        eqn = Eq(u.forward, stencil)
+        src_add = src.point2grid(u, m, u_t=t, p_t=time)
+        rec_read = Eq(rec, rec.grid2point(u))
+        stencils = [eqn] + src_add + [rec_read]
+
+        # TODO: The following time-index hackery is a legacy hangover
+        # from the Operator/Propagator structure and is used here for
+        # backward compatibiliy. We need re-examine this apporach carefully!
+
+        # Shift time indices so that LHS writes into t only,
+        # eg. u[t+2] = u[t+1] + u[t]  -> u[t] = u[t-1] + u[t-2]
+        stencils = [e.subs(t, t + solve(eqn.lhs.args[0], t)[0])
+                    if isinstance(e.lhs, TimeData) else e
+                    for e in stencils]
+        # Apply time substitutions as per legacy approach
+        time_subs = {t + 2: t + 1, t: t + 2, t - 2: t, t - 1: t + 1, t + 1: t}
```

Comment:
Aren't the above two operations self negating? 

What happens if you just leave both of them out?